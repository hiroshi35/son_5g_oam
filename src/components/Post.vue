<template>
  <article>
    <div class="panel panel-white article">
      <div class="panel-body">
        <div :class="'article-thumbnail ' + imgPosition" :style="thumbnail"></div>
        <h2>{{title}}</h2>
        <div>
          <small>by </small>
          <a href="#">{{author}} </a> |
          <small><i class="fa fa-clock-o" aria-hidden="true"></i> {{date}}</small>
        </div>
        <div class="article-body">
          <slot></slot>
        </div>
      </div>
      <div class="panel-footer">
        <a class="comments" href="#"><i class="fa fa-comments" aria-hidden="true"></i> {{commentsNumber}}</a> |
        <a @click.prevent="toggleLike" class="likes" :class="{'liked': likedByUser}" href="#"><i class="fa fa-heart-o"
                                                                                           aria-hidden="true"></i> {{likes}}</a>
        |
        <a href="#" class="share"><i class="fa fa-share-alt" aria-hidden="true"></i></a>
        <div class="category pull-right">
          <a href="#">{{category}}</a>
        </div>
      </div>
    </div>
  </article>
</template>

<script>
  export default {
    props: {
      author: {
        type: String
      },
      title: {
        type: String,
        required: true
      },
      date: {
        type: String
      },
      category: {
        type: String
      },
      commentsNumber: {
        type: Number
      },
      likesNumber: {
        type: Number
      },
      img: {
        type: String
      },
      imgPosition: {
        type: String,
        default: 'top'
      },
      liked: {
        type: Boolean,
        default: false
      }
    },
    data () {
      return {
        likes: this.likesNumber,
        likedByUser: this.liked
      }
    },
    computed: {
      thumbnail () {
        if (this.img) {
          if (this.img) {
            return {
              backgroundImage: 'url(' + this.img + ')',
              color: '#fff',
              backgroundRepeat: 'no-repeat',
              backgroundSize: 'cover'
            }
          }
        }
      }
    },
    methods: {
      toggleLike () {
        // Your code here
        if (!this.likedByUser) {
          this.likedByUser = true
          this.likes++
        } else {
          this.likedByUser = false
          this.likes--
        }
      }
    }
  }
</script>


<style lang="scss" scoped>
  $red: #dc5e5a;
  $blue: #65a4da;
  $teal: #3ab39b;
  article {
    .article {
      background-color: #fff;
      .article-body {
        margin-top: 10px;
      }
      .top {
        height: 200px;
        margin-top: -15px;
        margin-right: -15px;
        margin-left: -15px;
        border-top-left-radius: 5px;
        border-top-right-radius: 5px;
      }
      .likes {
        color: rgba($red, .8) !important;
        &:hover {
          color: $red !important;
        }
      }
      .comments {
        color: rgba($blue, .8);
        &:hover {
          color: $blue;
        }
      }

      .share {
        color: rgba($teal, .8);
        &:hover {
          color: $teal;
        }
      }

      .liked {
        color: darken($red, 10%) !important;
        font-weight: 600;

      }
    }
  }
</style>
