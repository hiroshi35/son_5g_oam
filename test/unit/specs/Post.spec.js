import { shallow } from 'vue-test-utils'
import Post from '@/components/Post.vue'

describe('Post.vue', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(Post, {
      propsData: {
        title: 'Title',
        author: 'John Doe',
        date: '2 hours ago',
        category: 'WEB',
        commentsNumber: 5,
        likesNumber: 5,
        img: '/img.jpg',
        imgPosition: 'top',
        liked: false
      },
      slots: {
        default: 'Body'
      }
    })
  })
  it('should be vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy()
  })
  it('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  it('has correct props', () => {
    expect(wrapper.props().title).toBe('Title')
    expect(wrapper.props().author).toBe('John Doe')
    expect(wrapper.props().date).toBe('2 hours ago')
    expect(wrapper.props().category).toBe('WEB')
    expect(wrapper.props().commentsNumber).toBe(5)
    expect(wrapper.props().likesNumber).toBe(5)
    expect(wrapper.props().img).toBe('/img.jpg')
    expect(wrapper.props().imgPosition).toBe('top')
    expect(wrapper.props().liked).toBe(false)
  })
  it('has correct data', () => {
    expect(wrapper.vm.likedByUser).toBe(false)
    expect(wrapper.vm.likes).toBe(5)
  })
  it('should increment likes number', () => {
    const btn = wrapper.find('.likes')
    btn.trigger('click')
    expect(wrapper.vm.likes).toBe(6)
    expect(wrapper.vm.likedByUser).toBe(true)
    expect(btn.classes()).toContain('liked')
  })
  it('should decrement likes number', () => {
    const btn = wrapper.find('.likes')
    btn.trigger('click')
    btn.trigger('click')
    expect(wrapper.vm.likedByUser).toBe(false)
    expect(wrapper.vm.likes).toBe(5)
    expect(btn.classes()).not.toContain('liked')
  })
})
