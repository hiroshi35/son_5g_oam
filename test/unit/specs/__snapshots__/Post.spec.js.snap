// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post.vue has the expected html structure 1`] = `
<article>
  <div
    class="panel panel-white article"
  >
    <div
      class="panel-body"
    >
      <div
        class="article-thumbnail top"
        style="background-image: url(/img.jpg); color: rgb(255, 255, 255); background-repeat: no-repeat; background-size: cover;"
      />
       
      <h2>
        Title
      </h2>
       
      <div>
        <small>
          by 
        </small>
         
        <a
          href="#"
        >
          John Doe 
        </a>
         |
        
        <small>
          <i
            aria-hidden="true"
            class="fa fa-clock-o"
          />
           2 hours ago
        </small>
      </div>
       
      <div
        class="article-body"
      >
        Body
      </div>
    </div>
     
    <div
      class="panel-footer"
    >
      <a
        class="comments"
        href="#"
      >
        <i
          aria-hidden="true"
          class="fa fa-comments"
        />
         5
      </a>
       |
      
      <a
        class="likes"
        href="#"
      >
        <i
          aria-hidden="true"
          class="fa fa-heart-o"
        />
         5
      </a>
      
      |
      
      <a
        class="share"
        href="#"
      >
        <i
          aria-hidden="true"
          class="fa fa-share-alt"
        />
      </a>
       
      <div
        class="category pull-right"
      >
        <a
          href="#"
        >
          WEB
        </a>
      </div>
    </div>
  </div>
</article>
`;
